tests/integration/awslambda/test_lambda_integration.py

```diff
@@ -450,77 +699,144 @@ def test_kinesis_lambda_parallelism(
     ):
         function_name = f"lambda_func-{short_uid()}"
         stream_name = f"test-foobar-{short_uid()}"
+        num_records_per_batch = 10
 
         create_lambda_function(
-            handler_file=TEST_LAMBDA_PARALLEL_FILE,
+            handler_file=TEST_LAMBDA_PYTHON_ECHO,
             func_name=function_name,
             runtime=LAMBDA_RUNTIME_PYTHON36,
             role=lambda_su_role,
         )
-
         kinesis_create_stream(StreamName=stream_name, ShardCount=1)
         stream_arn = kinesis_client.describe_stream(StreamName=stream_name)["StreamDescription"][
             "StreamARN"
         ]
-
         wait_for_stream_ready(stream_name=stream_name)
-
-        lambda_client.create_event_source_mapping(
+        event_source_uuid = lambda_client.create_event_source_mapping(
             EventSourceArn=stream_arn,
             FunctionName=function_name,
-            StartingPosition="TRIM_HORIZON",
-            BatchSize=10,
-        )
+            StartingPosition="LATEST",
+            BatchSize=num_records_per_batch,
+        )["UUID"]
+        _await_event_source_mapping_enabled(lambda_client, event_source_uuid)
 
-        stream_summary = kinesis_client.describe_stream_summary(StreamName=stream_name)
-        assert 1 == stream_summary["StreamDescriptionSummary"]["OpenShardCount"]
-        num_events_kinesis = 10
-        # assure async call
-        start = time.perf_counter()
         kinesis_client.put_records(
             Records=[
-                {"Data": '{"batch": 0}', "PartitionKey": f"test_{i}"}
-                for i in range(0, num_events_kinesis)
+                {"Data": json.dumps({"record_id": i}), "PartitionKey": "test"}
+                for i in range(0, num_records_per_batch)
             ],
             StreamName=stream_name,
         )
-        assert (time.perf_counter() - start) < 1  # this should not take more than a second
+
+        events = _get_lambda_invocation_events(logs_client, function_name, expected_num_events=1)
+        assert len(events) == 1
+
+        lambda_client.update_event_source_mapping(UUID=event_source_uuid, Enabled=False)
+        time.sleep(2)
         kinesis_client.put_records(
             Records=[
-                {"Data": '{"batch": 1}', "PartitionKey": f"test_{i}"}
-                for i in range(0, num_events_kinesis)
+                {"Data": json.dumps({"record_id": i}), "PartitionKey": "test"}
+                for i in range(0, num_records_per_batch)
             ],
             StreamName=stream_name,
         )
+        time.sleep(7)  # wait for records to pass through stream
```

Comment:
We'll need a more robust way here without hardcoding sleep values :thinking:

Reply:
Once I do the intended refactor I've mentioned in other comments here to add some kind of EventSourceListenerManager class it will be possible to accurately tell if event source mappings are enabled/disabled, so we'll be able to get rid of some of the hardcoded `sleep`s then. Currently it's not possible to do this due to the very loose, async coupling between the event source mapping API and the event source listener threads.

As for the second `sleep(7)` where we wait for records to pass through the stream before we make sure the lambda hasn't fired, I can't think of a great way to get around that (although I agree it's not a kludge). We have to wait for a while to make sure something _doesn't_ happen 🤷